dfc/keepalivetracker.go

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION. All rights reserved.
+ */
+
+package dfc
+
+// This file has different implementations of KeepaliveTracker.
+// ideally this file stays under dfcpub, but currently dfc has a flat structure, following that
+// for now.
+import (
+	"time"
+)
+
+// HeartBeatTracker tracks timestamp of the last a message is received from a server.
+// Timeout: a message is not received within the interval.
+type HeartBeatTracker struct {
+	ch       chan struct{}
+	last     map[string]time.Time
+	interval time.Duration // expected to hear from the server within the interval
+}
+
+// NewHeartBeatTracker returns a HeartBeatTracker
+func NewHeartBeatTracker(interval time.Duration) *HeartBeatTracker {
+	hb := &HeartBeatTracker{last: make(map[string]time.Time), ch: make(chan struct{}, 1)}
+	hb.interval = interval
+	hb.unlock()
+	return hb
+}
+
+func (hb *HeartBeatTracker) lock() {
+	<-hb.ch
+}
+
+func (hb *HeartBeatTracker) unlock() {
+	hb.ch <- struct{}{}
+}
+
+// HeardFrom is called to indicate a keepalive message (or equivalvent) is received from a server
+func (hb *HeartBeatTracker) HeardFrom(id string, reset bool) {
+	hb.lock()
+	defer hb.unlock()
+
+	hb.last[id] = time.Now()
+}
+
+// TimedOut returns true if it is determined that have not heard from the server
+func (hb *HeartBeatTracker) TimedOut(id string) bool {
+	hb.lock()
+	defer hb.unlock()
+
+	t, ok := hb.last[id]
+	return !ok || time.Since(t) > hb.interval
+}
+
+// AverageTracker keeps track of the average latency of all messages.
+// Timeout: last received is more than the 'factor' of current average.
+type AverageTracker struct {
+	ch     chan struct{}
+	rec    map[string]averageTrackerRecord
+	factor int
+}
+
+type averageTrackerRecord struct {
+	cnt     int64
+	last    time.Time
+	totalMS int64 // in ms
+}
+
+func (rec *averageTrackerRecord) avg() int64 {
+	return rec.totalMS / rec.cnt
+}
+
+// NewAverageTracker returns a AcerageTracker
```

Comment:
AcerageTracker -> AverageTracker

Reply:
done, thanks

Reply:
Can rec.cnt be null?

Reply:
yes, but avg() is only called after it checked rec.cnt == 0
that's always the debate of how much to check.